// @jsonforms/core/dist/test/src/util/field.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/*
  The MIT License

  Copyright (c) 2018 EclipseSource Munich
  https://github.com/eclipsesource/jsonforms

  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:

  The above copyright notice and this permission notice shall be included in
  all copies or substantial portions of the Software.

  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
  THE SOFTWARE.
*/
const isEmpty_1 = __importDefault(require("lodash/isEmpty"));
const has_1 = __importDefault(require("lodash/has"));
const cloneDeep_1 = __importDefault(require("lodash/cloneDeep"));
const merge_1 = __importDefault(require("lodash/merge"));
const reducers_1 = require("../reducers");
const util_1 = require("../util");
const renderer_1 = require("./renderer");
/**
 * Map state to field props.
 *
 * @param state JSONForms state tree
 * @param ownProps any own props
 * @returns {StatePropsOfField} state props of a field
 */
exports.mapStateToFieldProps = (state, ownProps) => {
    const path = util_1.composeWithUi(ownProps.uischema, ownProps.path);
    const visible = has_1.default(ownProps, 'visible')
        ? ownProps.visible
        : util_1.isVisible(ownProps, state);
    const enabled = has_1.default(ownProps, 'enabled')
        ? ownProps.enabled
        : util_1.isEnabled(ownProps, state);
    const errors = reducers_1.getErrorAt(path)(state).map(error => error.message);
    const isValid = isEmpty_1.default(errors);
    const controlElement = ownProps.uischema;
    const id = ownProps.id;
    const defaultConfig = cloneDeep_1.default(reducers_1.getConfig(state));
    const config = merge_1.default(defaultConfig, ownProps.uischema.options);
    return {
        data: ownProps.data !== undefined
            ? util_1.Resolve.data(ownProps.data, path)
            : util_1.Resolve.data(reducers_1.getData(state), path),
        visible,
        enabled,
        id,
        path,
        errors,
        isValid,
        scopedSchema: util_1.Resolve.schema(ownProps.schema, controlElement.scope),
        uischema: ownProps.uischema,
        schema: ownProps.schema,
        config,
        findUISchema: reducers_1.findUISchema(state)
    };
};
exports.mapStateToDispatchFieldProps = (state, ownProps) => {
    const props = exports.mapStateToFieldProps(state, ownProps);
    return Object.assign({}, props, ownProps, { fields: state.jsonforms.fields || [] });
};
/**
 * Default mapStateToFieldProps for enum field. Options is used for populating dropdown list
 * @param state
 * @param ownProps
 * @returns {StatePropsOfEnumField}
 */
exports.defaultMapStateToEnumFieldProps = (state, ownProps) => {
    const props = exports.mapStateToFieldProps(state, ownProps);
    const options = ownProps.options !== undefined ? ownProps.options : props.scopedSchema.enum;
    return Object.assign({}, props, { options });
};
/**
 * Synonym for mapDispatchToControlProps.
 *
 * @type {(dispatch) => {handleChange(path, value): void}}
 */
exports.mapDispatchToFieldProps = renderer_1.mapDispatchToControlProps;
/**
 * Default dispatch to control props which can be customized to set handleChange action
 *
 */
exports.defaultMapDispatchToControlProps = 
// TODO: ownProps types
(dispatch, ownProps) => {
    const dispatchControlProps = exports.mapDispatchToFieldProps(dispatch);
    return {
        handleChange: ownProps.handleChange !== undefined
            ? ownProps.handleChange
            : dispatchControlProps.handleChange
    };
};
