// @jsonforms/core/dist/test/src/util/renderer.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/*
  The MIT License
  
  Copyright (c) 2018 EclipseSource Munich
  https://github.com/eclipsesource/jsonforms
  
  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:
  
  The above copyright notice and this permission notice shall be included in
  all copies or substantial portions of the Software.
  
  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
  THE SOFTWARE.
*/
const get_1 = __importDefault(require("lodash/get"));
const has_1 = __importDefault(require("lodash/has"));
const cloneDeep_1 = __importDefault(require("lodash/cloneDeep"));
const merge_1 = __importDefault(require("lodash/merge"));
const union_1 = __importDefault(require("lodash/union"));
const reducers_1 = require("../reducers");
const util_1 = require("../util");
const actions_1 = require("../actions");
const generators_1 = require("../generators");
exports.isPlainLabel = (label) => {
    return typeof label === 'string';
};
const isRequired = (schema, schemaPath) => {
    const pathSegments = schemaPath.split('/');
    const lastSegment = pathSegments[pathSegments.length - 1];
    const nextHigherSchemaSegments = pathSegments.slice(0, pathSegments.length - 2);
    const nextHigherSchemaPath = nextHigherSchemaSegments.join('/');
    const nextHigherSchema = util_1.Resolve.schema(schema, nextHigherSchemaPath);
    return (nextHigherSchema !== undefined &&
        nextHigherSchema.required !== undefined &&
        nextHigherSchema.required.indexOf(lastSegment) !== -1);
};
/**
 * Adds an asterisk to the given label string based
 * on the required parameter.
 *
 * @param {string} label the label string
 * @param {boolean} required whether the label belongs to a control which is required
 * @returns {string} the label string
 */
exports.computeLabel = (label, required) => {
    return required ? label + '*' : label;
};
/**
 * Create a default value based on the given scheam.
 * @param schema the schema for which to create a default value.
 * @returns {any}
 */
exports.createDefaultValue = (schema) => {
    switch (schema.type) {
        case 'string':
            if (schema.format === 'date-time' ||
                schema.format === 'date' ||
                schema.format === 'time') {
                return new Date();
            }
            return '';
        case 'integer':
        case 'number':
            return 0;
        case 'boolean':
            return false;
        case 'array':
            return [];
        case 'null':
            return null;
        default:
            return {};
    }
};
/**
 * Whether an element's description should be hidden.
 *
 * @param visible whether an element is visible
 * @param description the element's description
 * @param isFocused whether the element is focused
 *
 * @returns {boolean} true, if the description is to be hidden, false otherwise
 */
exports.isDescriptionHidden = (visible, description, isFocused) => {
    return (description === undefined ||
        (description !== undefined && !visible) ||
        !isFocused);
};
/**
 * Map state to control props.
 * @param state the store's state
 * @param ownProps any own props
 * @returns {StatePropsOfControl} state props for a control
 */
exports.mapStateToControlProps = (state, ownProps) => {
    const { uischema } = ownProps;
    const path = util_1.composeWithUi(uischema, ownProps.path);
    const visible = has_1.default(ownProps, 'visible')
        ? ownProps.visible
        : util_1.isVisible(ownProps, state, ownProps.path);
    const enabled = has_1.default(ownProps, 'enabled')
        ? ownProps.enabled
        : util_1.isEnabled(ownProps, state, ownProps.path);
    const labelDesc = util_1.createLabelDescriptionFrom(uischema);
    const label = labelDesc.show ? labelDesc.text : '';
    const errors = union_1.default(reducers_1.getErrorAt(path)(state).map(error => error.message));
    const controlElement = uischema;
    const id = ownProps.id;
    const required = controlElement.scope !== undefined &&
        isRequired(ownProps.schema, controlElement.scope);
    const resolvedSchema = util_1.Resolve.schema(ownProps.schema, controlElement.scope);
    const description = resolvedSchema !== undefined ? resolvedSchema.description : '';
    const defaultConfig = cloneDeep_1.default(reducers_1.getConfig(state));
    const config = merge_1.default(defaultConfig, controlElement.options);
    return {
        data: util_1.Resolve.data(reducers_1.getData(state), path),
        description,
        errors,
        label,
        visible,
        enabled,
        id,
        path,
        parentPath: ownProps.path,
        required,
        scopedSchema: resolvedSchema,
        uischema: ownProps.uischema,
        findUISchema: reducers_1.findUISchema(state),
        schema: ownProps.schema,
        config,
        fields: state.jsonforms.fields
    };
};
/**
 *
 * Map dispatch to control props.
 *
 * @param dispatch the store's dispatch method
 * @returns {DispatchPropsOfControl} dispatch props for a control
 */
exports.mapDispatchToControlProps = (dispatch) => ({
    handleChange(path, value) {
        dispatch(actions_1.update(path, () => value));
    }
});
/**
 * Map state to table props
 *
 * @param state the store's state
 * @param ownProps any element's own props
 * @returns {StatePropsOfArrayControl} state props for a table control
 */
exports.mapStateToArrayControlProps = (state, ownProps) => {
    const _a = exports.mapStateToControlProps(state, ownProps), { path, schema, uischema } = _a, props = __rest(_a, ["path", "schema", "uischema"]);
    const controlElement = uischema;
    const resolvedSchema = util_1.Resolve.schema(schema, controlElement.scope + '/items');
    const childErrors = reducers_1.getSubErrorsAt(path)(state);
    return Object.assign({}, props, { path,
        schema,
        uischema, scopedSchema: resolvedSchema, childErrors });
};
/**
 * Maps state to dispatch properties of an array control.
 *
 * @param dispatch the store's dispatch method
 * @param ownProps own properties
 * @returns {DispatchPropsOfArrayControl} dispatch props of an array control
 */
exports.mapDispatchToArrayControlProps = (dispatch, ownProps) => ({
    addItem: (path) => () => {
        dispatch(actions_1.update(path, array => {
            const schemaPath = ownProps.uischema.scope + '/items';
            const resolvedSchema = util_1.Resolve.schema(ownProps.schema, schemaPath);
            const newValue = exports.createDefaultValue(resolvedSchema);
            if (array === undefined || array === null) {
                return [newValue];
            }
            array.push(newValue);
            return array;
        }));
    },
    removeItems: (path, toDelete) => () => {
        dispatch(actions_1.update(path, array => {
            toDelete.forEach(s => array.splice(array.indexOf(s), 1));
            return array;
        }));
    }
});
/**
 * Map state to layout props.
 * @param state JSONForms state tree
 * @param ownProps any own props
 * @returns {StatePropsOfLayout}
 */
exports.mapStateToLayoutProps = (state, ownProps) => {
    const visible = has_1.default(ownProps, 'visible')
        ? ownProps.visible
        : util_1.isVisible(ownProps, state, ownProps.path);
    return {
        renderers: reducers_1.getRenderers(state),
        visible,
        path: ownProps.path,
        uischema: ownProps.uischema,
        schema: ownProps.schema
    };
};
exports.mapStateToJsonFormsRendererProps = (state, ownProps) => {
    let uischema = ownProps.uischema;
    if (uischema === undefined) {
        if (ownProps.schema) {
            uischema = generators_1.generateDefaultUISchema(ownProps.schema);
        }
        else {
            uischema = reducers_1.getUiSchema(state);
        }
    }
    return {
        renderers: ownProps.renderers || get_1.default(state.jsonforms, 'renderers') || [],
        schema: ownProps.schema || reducers_1.getSchema(state),
        uischema
    };
};
