// @jsonforms/ionic-renderers/lib/json-forms.module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var ionic_angular_1 = require("ionic-angular");
var ionic_selectable_1 = require("ionic-selectable");
var angular_1 = require("@jsonforms/angular");
var angular_l10n_1 = require("angular-l10n");
var boolean_checkbox_control_1 = require("./controls/boolean/boolean-checkbox-control");
var boolean_toggle_control_1 = require("./controls/boolean/boolean-toggle-control");
var string_control_1 = require("./controls/string/string-control");
var multiline_control_1 = require("./controls/string/multiline-control");
var number_control_1 = require("./controls/number/number-control");
var date_control_1 = require("./controls/date/date-control");
var enum_control_1 = require("./controls/enum/enum-control");
var range_control_1 = require("./controls/range/range-control");
var horizontal_layout_1 = require("./layouts/horizontal/horizontal-layout");
var vertical_layout_1 = require("./layouts/vertical/vertical-layout");
var categorization_menu_layout_1 = require("./layouts/categorization/categorization-menu-layout");
var category_1 = require("./layouts/categorization/category/category");
var group_layout_1 = require("./layouts/group/group-layout");
var list_with_detail_control_1 = require("./other/list-with-detail/list-with-detail-control");
var master_1 = require("./other/list-with-detail/pages/master/master");
var detail_1 = require("./other/list-with-detail/pages/detail/detail");
var label_1 = require("./other/label/label");
var categorization_tab_layout_1 = require("./layouts/categorization/categorization-tab-layout");
var autocomplete_control_1 = require("./controls/enum/autocomplete-control");
var object_control_1 = require("./controls/object/object.control");
var ParamsService_1 = require("./services/ParamsService");
var emptyL10NConfig = {};
var JsonFormsIonicModule = /** @class */ (function () {
    function JsonFormsIonicModule() {
    }
    JsonFormsIonicModule = __decorate([
        core_1.NgModule({
            declarations: [
                // controls
                boolean_checkbox_control_1.BooleanCheckboxControlRenderer,
                boolean_toggle_control_1.BooleanToggleControlRenderer,
                string_control_1.StringControlRenderer,
                multiline_control_1.MultilineControlRenderer,
                number_control_1.NumberControlRenderer,
                date_control_1.DateControlRenderer,
                enum_control_1.EnumControlRenderer,
                range_control_1.RangeControlRenderer,
                autocomplete_control_1.AutoCompleteControlRenderer,
                object_control_1.ObjectControlRenderer,
                // layouts
                horizontal_layout_1.HorizontalLayoutRenderer,
                vertical_layout_1.VerticalLayoutRenderer,
                categorization_tab_layout_1.CategorizationTabLayoutRenderer,
                categorization_menu_layout_1.CategorizationMenuLayoutRenderer,
                category_1.CategoryRenderer,
                group_layout_1.GroupLayoutRenderer,
                // ListWithDetail components
                list_with_detail_control_1.ListWithDetailControl,
                master_1.MasterPage,
                detail_1.DetailPage,
                // other
                label_1.LabelRenderer
            ],
            imports: [
                ionic_angular_1.IonicModule,
                ionic_selectable_1.IonicSelectableModule,
                angular_1.JsonFormsModule,
                angular_l10n_1.LocalizationModule,
                angular_l10n_1.LocaleValidationModule.forRoot(),
                angular_l10n_1.TranslationModule.forRoot(emptyL10NConfig)
            ],
            exports: [
                ionic_angular_1.IonicModule,
                ionic_selectable_1.IonicSelectableModule,
                angular_1.JsonFormsModule,
                angular_l10n_1.LocalizationModule,
                angular_l10n_1.LocaleValidationModule,
                angular_l10n_1.TranslationModule
            ],
            entryComponents: [
                // controls
                boolean_checkbox_control_1.BooleanCheckboxControlRenderer,
                boolean_toggle_control_1.BooleanToggleControlRenderer,
                string_control_1.StringControlRenderer,
                multiline_control_1.MultilineControlRenderer,
                number_control_1.NumberControlRenderer,
                date_control_1.DateControlRenderer,
                enum_control_1.EnumControlRenderer,
                range_control_1.RangeControlRenderer,
                autocomplete_control_1.AutoCompleteControlRenderer,
                object_control_1.ObjectControlRenderer,
                // layouts
                horizontal_layout_1.HorizontalLayoutRenderer,
                vertical_layout_1.VerticalLayoutRenderer,
                categorization_menu_layout_1.CategorizationMenuLayoutRenderer,
                categorization_tab_layout_1.CategorizationTabLayoutRenderer,
                category_1.CategoryRenderer,
                group_layout_1.GroupLayoutRenderer,
                // ListWithDetail components
                list_with_detail_control_1.ListWithDetailControl,
                master_1.MasterPage,
                detail_1.DetailPage,
                // other
                label_1.LabelRenderer
            ],
            providers: [ParamsService_1.ParamsService]
        })
    ], JsonFormsIonicModule);
    return JsonFormsIonicModule;
}());
exports.JsonFormsIonicModule = JsonFormsIonicModule;
