// @jsonforms/core/dist/test/src/actions/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const generators_1 = require("../generators");
exports.INIT = 'jsonforms/INIT';
exports.SET_AJV = 'jsonforms/SET_AJV';
exports.UPDATE_DATA = 'jsonforms/UPDATE';
exports.VALIDATE = 'jsonforms/VALIDATE';
exports.ADD_RENDERER = 'jsonforms/ADD_RENDERER';
exports.REMOVE_RENDERER = 'jsonforms/REMOVE_RENDERER';
exports.ADD_FIELD = 'jsonforms/ADD_FIELD';
exports.REMOVE_FIELD = 'jsonforms/REMOVE_FIELD';
exports.SET_CONFIG = 'jsonforms/SET_CONFIG';
exports.ADD_UI_SCHEMA = `jsonforms/ADD_UI_SCHEMA`;
exports.REMOVE_UI_SCHEMA = `jsonforms/REMOVE_UI_SCHEMA`;
exports.SET_SCHEMA = `jsonforms/SET_SCHEMA`;
exports.SET_UISCHEMA = `jsonforms/SET_UISCHEMA`;
exports.SET_LOCALE = `jsonforms/SET_LOCALE`;
exports.SET_LOCALIZED_SCHEMAS = 'jsonforms/SET_LOCALIZED_SCHEMAS';
exports.SET_LOCALIZED_UISCHEMAS = 'jsonforms/SET_LOCALIZED_UISCHEMAS';
exports.ADD_DEFAULT_DATA = `jsonforms/ADD_DEFAULT_DATA`;
exports.REMOVE_DEFAULT_DATA = `jsonforms/REMOVE_DEFAULT_DATA`;
exports.init = (data, schema = generators_1.generateJsonSchema(data), uischema = generators_1.generateDefaultUISchema(schema), ajv) => ({
    type: exports.INIT,
    data,
    schema,
    uischema,
    ajv
});
exports.registerDefaultData = (schemaPath, data) => ({
    type: exports.ADD_DEFAULT_DATA,
    schemaPath,
    data
});
exports.unregisterDefaultData = (schemaPath) => ({
    type: exports.REMOVE_DEFAULT_DATA,
    schemaPath
});
exports.setAjv = (ajv) => ({
    type: exports.SET_AJV,
    ajv
});
exports.update = (path, updater) => ({
    type: exports.UPDATE_DATA,
    path,
    updater
});
exports.registerRenderer = (tester, renderer) => ({
    type: exports.ADD_RENDERER,
    tester,
    renderer
});
exports.registerField = (tester, field) => ({
    type: exports.ADD_FIELD,
    tester,
    field
});
exports.unregisterField = (tester, field) => ({
    type: exports.REMOVE_FIELD,
    tester,
    field
});
exports.unregisterRenderer = (tester, renderer) => ({
    type: exports.REMOVE_RENDERER,
    tester,
    renderer
});
exports.setConfig = (config) => (dispatch) => {
    dispatch({
        type: exports.SET_CONFIG,
        config
    });
};
exports.registerUISchema = (tester, uischema) => {
    return {
        type: exports.ADD_UI_SCHEMA,
        tester,
        uischema
    };
};
exports.unregisterUISchema = (tester) => {
    return {
        type: exports.REMOVE_UI_SCHEMA,
        tester
    };
};
exports.setLocale = (locale) => ({
    type: exports.SET_LOCALE,
    locale
});
exports.setLocalizedSchemas = (localizedSchemas) => ({
    type: exports.SET_LOCALIZED_SCHEMAS,
    localizedSchemas
});
exports.setSchema = (schema) => ({
    type: exports.SET_SCHEMA,
    schema
});
exports.setLocalizedUISchemas = (localizedUISchemas) => ({
    type: exports.SET_LOCALIZED_UISCHEMAS,
    localizedUISchemas
});
exports.setUISchema = (uischema) => ({
    type: exports.SET_UISCHEMA,
    uischema
});
