// @jsonforms/core/dist/test/src/reducers/core.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/*
  The MIT License
  
  Copyright (c) 2018 EclipseSource Munich
  https://github.com/eclipsesource/jsonforms
  
  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:
  
  The above copyright notice and this permission notice shall be included in
  all copies or substantial portions of the Software.
  
  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
  THE SOFTWARE.
*/
const cloneDeep_1 = __importDefault(require("lodash/cloneDeep"));
const set_1 = __importDefault(require("lodash/set"));
const get_1 = __importDefault(require("lodash/get"));
const filter_1 = __importDefault(require("lodash/filter"));
const actions_1 = require("../actions");
const validator_1 = require("../util/validator");
const validate = (validator, data) => {
    const valid = validator(data);
    if (valid) {
        return [];
    }
    return validator.errors;
};
const sanitizeErrors = (validator, data) => validate(validator, data).map(error => {
    error.dataPath = error.dataPath.replace(/\//g, '.').substr(1);
    return error;
});
const alwaysValid = () => true;
const initState = {
    data: {},
    schema: {},
    uischema: undefined,
    errors: [],
    validator: alwaysValid,
    ajv: undefined
};
const getOrCreateAjv = (state, action) => {
    if (action && action.ajv) {
        return action.ajv;
    }
    if (state.ajv) {
        return state.ajv;
    }
    return validator_1.createAjv();
};
exports.coreReducer = (state = initState, action) => {
    switch (action.type) {
        case actions_1.INIT: {
            const thisAjv = getOrCreateAjv(state, action);
            const v = thisAjv.compile(action.schema);
            const e = sanitizeErrors(v, action.data);
            return Object.assign({}, state, { data: action.data, schema: action.schema, uischema: action.uischema, errors: e, validator: v, ajv: thisAjv });
        }
        case actions_1.SET_AJV: {
            const currentAjv = action.ajv;
            const validator = currentAjv.compile(state.schema);
            const errors = sanitizeErrors(validator, state.data);
            return Object.assign({}, state, { validator,
                errors });
        }
        case actions_1.SET_SCHEMA: {
            const v = action.schema && state.ajv
                ? state.ajv.compile(action.schema)
                : state.validator;
            return Object.assign({}, state, { validator: v, schema: action.schema });
        }
        case actions_1.SET_UISCHEMA: {
            return Object.assign({}, state, { uischema: action.uischema });
        }
        case actions_1.UPDATE_DATA: {
            if (action.path === undefined || action.path === null) {
                return state;
            }
            else if (action.path === '') {
                // empty path is ok
                const result = action.updater(cloneDeep_1.default(state.data));
                if (result === undefined || result === null) {
                    return Object.assign({}, state, { data: state.data, uischema: state.uischema, schema: state.schema });
                }
                const errors = sanitizeErrors(state.validator, result);
                return Object.assign({}, state, { data: result, errors });
            }
            else {
                const oldData = get_1.default(state.data, action.path);
                let newData = action.updater(oldData);
                if (newData === '') {
                    newData = undefined;
                }
                const newState = set_1.default(cloneDeep_1.default(state.data), action.path, newData);
                const errors = sanitizeErrors(state.validator, newState);
                return Object.assign({}, state, { data: newState, errors });
            }
        }
        default:
            return state;
    }
};
exports.extractData = (state) => get_1.default(state, 'data');
exports.extractSchema = (state) => get_1.default(state, 'schema');
exports.extractUiSchema = (state) => get_1.default(state, 'uischema');
exports.errorAt = (instancePath) => (state) => {
    return filter_1.default(state.errors, (error) => error.dataPath === instancePath);
};
exports.subErrorsAt = (instancePath) => (state) => {
    const path = `${instancePath}.`;
    return filter_1.default(state.errors, (error) => error.dataPath.startsWith(path));
};
