// @jsonforms/core/dist/test/src/reducers/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/*
  The MIT License
  
  Copyright (c) 2018 EclipseSource Munich
  https://github.com/eclipsesource/jsonforms
  
  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:
  
  The above copyright notice and this permission notice shall be included in
  all copies or substantial portions of the Software.
  
  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
  THE SOFTWARE.
*/
const get_1 = __importDefault(require("lodash/get"));
const default_data_1 = require("./default-data");
const redux_1 = require("redux");
const renderers_1 = require("./renderers");
exports.rendererReducer = renderers_1.rendererReducer;
const fields_1 = require("./fields");
exports.fieldReducer = fields_1.fieldReducer;
const config_1 = require("./config");
const core_1 = require("./core");
exports.coreReducer = core_1.coreReducer;
const uischemas_1 = require("./uischemas");
const __1 = require("..");
const i18n_1 = require("./i18n");
exports.jsonformsReducer = (additionalReducers = {}) => redux_1.combineReducers(Object.assign({ core: core_1.coreReducer, renderers: renderers_1.rendererReducer, fields: fields_1.fieldReducer, config: config_1.configReducer, uischemas: uischemas_1.uischemaRegistryReducer, defaultData: default_data_1.defaultDataReducer, i18n: i18n_1.i18nReducer }, additionalReducers));
exports.getData = (state) => core_1.extractData(get_1.default(state, 'jsonforms.core'));
exports.getSchema = (state) => core_1.extractSchema(get_1.default(state, 'jsonforms.core'));
exports.getUiSchema = (state) => core_1.extractUiSchema(get_1.default(state, 'jsonforms.core'));
exports.getDefaultData = (state) => default_data_1.extractDefaultData(get_1.default(state, 'jsonforms.defaultData'));
exports.getRenderers = (state) => get_1.default(state, 'jsonforms.renderers');
exports.findUISchema = (state) => (schema, schemaPath, path, fallbackLayoutType = 'VerticalLayout', control) => {
    // handle options
    if (control && control.options && control.options.detail) {
        if (typeof control.options.detail === 'string') {
            if (control.options.detail.toUpperCase() === 'GENERATE') {
                // force generation of uischema
                return __1.Generate.uiSchema(schema, fallbackLayoutType);
            }
        }
        else if (typeof control.options.detail === 'object') {
            // check if detail is a valid uischema
            if (control.options.detail.type &&
                typeof control.options.detail.type === 'string') {
                return control.options.detail;
            }
        }
    }
    // default
    const uiSchema = uischemas_1.findMatchingUISchema(state.jsonforms.uischemas)(schema, schemaPath, path);
    if (uiSchema === undefined) {
        return __1.Generate.uiSchema(schema, fallbackLayoutType);
    }
    return uiSchema;
};
exports.getErrorAt = (instancePath) => (state) => {
    return core_1.errorAt(instancePath)(state.jsonforms.core);
};
exports.getSubErrorsAt = (instancePath) => (state) => core_1.subErrorsAt(instancePath)(state.jsonforms.core);
exports.getConfig = (state) => state.jsonforms.config;
exports.getLocale = (state) => i18n_1.fetchLocale(get_1.default(state, 'jsonforms.i18n'));
exports.getLocalizedSchema = (locale) => (state) => i18n_1.findLocalizedSchema(locale)(get_1.default(state, 'jsonforms.i18n'));
exports.getLocalizedUISchema = (locale) => (state) => i18n_1.findLocalizedUISchema(locale)(get_1.default(state, 'jsonforms.i18n'));
