// @jsonforms/core/dist/test/test/util/field.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
/*
  The MIT License

  Copyright (c) 2018 EclipseSource Munich
  https://github.com/eclipsesource/jsonforms

  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:

  The above copyright notice and this permission notice shall be included in
  all copies or substantial portions of the Software.

  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
  THE SOFTWARE.
*/
const ava_1 = __importDefault(require("ava"));
const _ = __importStar(require("lodash"));
const util_1 = require("../../src/util");
const actions_1 = require("../../src/actions");
const redux_mock_store_1 = __importDefault(require("redux-mock-store"));
const src_1 = require("../../src");
const middlewares = [];
const mockStore = redux_mock_store_1.default(middlewares);
const hideRule = {
    effect: src_1.RuleEffect.HIDE,
    condition: {
        type: 'LEAF',
        scope: '#/properties/firstName',
        expectedValue: 'Homer'
    }
};
const disableRule = {
    effect: src_1.RuleEffect.DISABLE,
    condition: {
        type: 'LEAF',
        scope: '#/properties/firstName',
        expectedValue: 'Homer'
    }
};
const coreUISchema = {
    type: 'Control',
    scope: '#/properties/firstName'
};
const createState = (uischema) => ({
    jsonforms: {
        core: {
            schema: {
                type: 'object',
                properties: {
                    firstName: { type: 'string' },
                    lastName: { type: 'string' },
                    nationality: {
                        type: 'string',
                        enum: ['DE', 'IT', 'JP', 'US', 'RU', 'Other']
                    }
                }
            },
            data: {
                firstName: 'Homer'
            },
            uischema,
            errors: []
        }
    }
});
ava_1.default('mapStateToFieldProps - visible via ownProps ', t => {
    const uischema = Object.assign({}, coreUISchema, { rule: hideRule });
    const ownProps = {
        visible: true,
        uischema
    };
    const props = util_1.mapStateToFieldProps(createState(uischema), ownProps);
    t.true(props.visible);
});
ava_1.default('mapStateToFieldProps - hidden via ownProps ', t => {
    const uischema = Object.assign({}, coreUISchema, { rule: hideRule });
    const ownProps = {
        visible: false,
        uischema
    };
    const props = util_1.mapStateToFieldProps(createState(uischema), ownProps);
    t.false(props.visible);
});
ava_1.default('mapStateToFieldProps - hidden via state ', t => {
    const uischema = Object.assign({}, coreUISchema, { rule: hideRule });
    const ownProps = {
        uischema
    };
    const props = util_1.mapStateToFieldProps(createState(uischema), ownProps);
    t.false(props.visible);
});
ava_1.default('mapStateToFieldProps - visible via state ', t => {
    const uischema = Object.assign({}, coreUISchema, { rule: hideRule });
    const ownProps = {
        uischema
    };
    const clonedState = _.cloneDeep(createState(uischema));
    clonedState.jsonforms.core.data.firstName = 'Lisa';
    const props = util_1.mapStateToFieldProps(clonedState, ownProps);
    t.true(props.visible);
});
ava_1.default('mapStateToFieldProps - enabled via ownProps ', t => {
    const uischema = Object.assign({}, coreUISchema, { rule: disableRule });
    const ownProps = {
        enabled: true,
        uischema
    };
    const props = util_1.mapStateToFieldProps(createState(uischema), ownProps);
    t.true(props.enabled);
});
ava_1.default('mapStateToFieldProps - disabled via ownProps ', t => {
    const uischema = Object.assign({}, coreUISchema, { rule: disableRule });
    const ownProps = {
        enabled: false,
        uischema
    };
    const props = util_1.mapStateToFieldProps(createState(uischema), ownProps);
    t.false(props.enabled);
});
ava_1.default('mapStateToFieldProps - disabled via state ', t => {
    const uischema = Object.assign({}, coreUISchema, { rule: disableRule });
    const ownProps = {
        uischema
    };
    const props = util_1.mapStateToFieldProps(createState(uischema), ownProps);
    t.false(props.enabled);
});
ava_1.default('mapStateToFieldProps - enabled via state ', t => {
    const uischema = Object.assign({}, coreUISchema, { rule: disableRule });
    const ownProps = {
        uischema
    };
    const clonedState = _.cloneDeep(createState(uischema));
    clonedState.jsonforms.core.data.firstName = 'Lisa';
    const props = util_1.mapStateToFieldProps(clonedState, ownProps);
    t.true(props.enabled);
});
ava_1.default('mapStateToFieldProps - path', t => {
    const ownProps = {
        uischema: coreUISchema
    };
    const props = util_1.mapStateToFieldProps(createState(coreUISchema), ownProps);
    t.is(props.path, 'firstName');
});
ava_1.default('mapStateToFieldProps - compose path with ownProps.path', t => {
    const ownProps = {
        uischema: coreUISchema,
        path: 'yo'
    };
    const props = util_1.mapStateToFieldProps(createState(coreUISchema), ownProps);
    t.is(props.path, 'yo.firstName');
});
ava_1.default('mapStateToFieldProps - data', t => {
    const ownProps = {
        uischema: coreUISchema
    };
    const props = util_1.mapStateToFieldProps(createState(coreUISchema), ownProps);
    t.is(props.data, 'Homer');
});
ava_1.default('mapStateToFieldProps - id', t => {
    util_1.clearAllIds();
    const ownProps = {
        uischema: coreUISchema,
        id: '#/properties/firstName'
    };
    const props = util_1.mapStateToFieldProps(createState(coreUISchema), ownProps);
    t.is(props.id, '#/properties/firstName');
});
ava_1.default('mapStateToEnumFieldProps - set default options for dropdown list', t => {
    const uiSchema = {
        type: 'Control',
        scope: '#/properties/nationality'
    };
    const ownProps = {
        schema: {
            type: 'object',
            properties: {
                firstName: { type: 'string' },
                lastName: { type: 'string' },
                nationality: {
                    type: 'string',
                    enum: ['DE', 'IT', 'JP', 'US', 'RU', 'Other']
                }
            }
        },
        uischema: uiSchema
    };
    const props = util_1.defaultMapStateToEnumFieldProps(createState(uiSchema), ownProps);
    t.deepEqual(props.options, ['DE', 'IT', 'JP', 'US', 'RU', 'Other']);
    t.is(props.data, undefined);
});
ava_1.default('mapStateToFieldProps - set data of enum field', t => {
    const uiSchema = {
        type: 'Control',
        scope: '#/properties/nationality'
    };
    const ownProps = {
        data: {
            nationality: 'JP'
        },
        schema: {
            type: 'object',
            properties: {
                firstName: { type: 'string' },
                lastName: { type: 'string' },
                nationality: {
                    type: 'string',
                    enum: ['DE', 'IT', 'JP', 'US', 'RU', 'Other']
                }
            }
        },
        uischema: uiSchema
    };
    const props = util_1.mapStateToFieldProps(createState(uiSchema), ownProps);
    t.is(props.data, 'JP');
});
ava_1.default('defaultMapDispatchToControlProps, initialized with custom handleChange', t => {
    let didChange = false;
    const uiSchema = {
        type: 'Control',
        scope: '#/properties/nationality'
    };
    const ownProps = {
        handleChange: () => {
            didChange = true;
        }
    };
    const store = mockStore(createState(uiSchema));
    const props = util_1.defaultMapDispatchToControlProps(store.dispatch, ownProps);
    props.handleChange(undefined, undefined);
    t.true(didChange);
});
ava_1.default('defaultMapDispatchToControlProps, with default handleChange', t => {
    const uiSchema = {
        type: 'Control',
        scope: '#/properties/nationality'
    };
    const store = mockStore(createState(uiSchema));
    const props = util_1.defaultMapDispatchToControlProps(store.dispatch, {});
    props.handleChange('nationality', 'DE');
    const updateAction = _.head(store.getActions());
    t.is(updateAction.type, actions_1.UPDATE_DATA);
    t.is(updateAction.path, 'nationality');
    t.is(updateAction.updater(), 'DE');
});
