// @jsonforms/core/dist/test/test/util/path.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/*
  The MIT License
  
  Copyright (c) 2018 EclipseSource Munich
  https://github.com/eclipsesource/jsonforms
  
  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:
  
  The above copyright notice and this permission notice shall be included in
  all copies or substantial portions of the Software.
  
  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
  THE SOFTWARE.
*/
const ava_1 = __importDefault(require("ava"));
const util_1 = require("../../src/util");
ava_1.default('resolve ', t => {
    const schema = {
        type: 'object',
        properties: {
            foo: {
                type: 'integer'
            }
        }
    };
    t.deepEqual(util_1.Resolve.schema(schema, '#/properties/foo'), {
        type: 'integer'
    });
});
ava_1.default('toDataPath ', t => {
    t.is(util_1.toDataPath('#/properties/foo/properties/bar'), 'foo.bar');
});
ava_1.default('toDataPath use of keywords', t => {
    t.is(util_1.toDataPath('#/properties/properties'), 'properties');
});
ava_1.default('toDataPath use of encoded paths', t => {
    const fooBar = encodeURIComponent('foo.bar');
    t.is(util_1.toDataPath(`#/properties/${fooBar}`), `${fooBar}`);
});
ava_1.default('toDataPath relative with /', t => {
    t.is(util_1.toDataPath('/properties/foo/properties/bar'), 'foo.bar');
});
ava_1.default('toDataPath use of keywords relative with /', t => {
    t.is(util_1.toDataPath('/properties/properties'), 'properties');
});
ava_1.default('toDataPath use of encoded paths relative with /', t => {
    const fooBar = encodeURIComponent('foo/bar');
    t.is(util_1.toDataPath(`/properties/${fooBar}`), `${fooBar}`);
});
ava_1.default('toDataPath relative without /', t => {
    t.is(util_1.toDataPath('properties/foo/properties/bar'), 'foo.bar');
});
ava_1.default('toDataPath use of keywords relative without /', t => {
    t.is(util_1.toDataPath('properties/properties'), 'properties');
});
ava_1.default('toDataPath use of encoded paths relative without /', t => {
    const fooBar = encodeURIComponent('foo/bar');
    t.is(util_1.toDataPath(`properties/${fooBar}`), `${fooBar}`);
});
ava_1.default('resolve instance', t => {
    const instance = { foo: 123 };
    const result = util_1.Resolve.data(instance, util_1.toDataPath('#/properties/foo'));
    t.is(result, 123);
});
ava_1.default('resolve instance with keywords', t => {
    const instance = { properties: 123 };
    const result = util_1.Resolve.data(instance, util_1.toDataPath('#/properties/properties'));
    t.is(result, 123);
});
ava_1.default('resolve instance with encoded', t => {
    const instance = { 'foo/bar': 123 };
    const fooBar = encodeURIComponent('foo/bar');
    const result = util_1.Resolve.data(instance, util_1.toDataPath(`#/properties/${fooBar}`));
    t.is(result, 123);
});
ava_1.default('resolve nested instance', t => {
    const instance = { foo: { bar: 123 } };
    const result = util_1.Resolve.data(instance, util_1.toDataPath('#/properties/foo/properties/bar'));
    t.is(result, 123);
});
ava_1.default('resolve uninitiated instance', t => {
    const instance = {};
    const result = util_1.Resolve.data(instance, util_1.toDataPath('#/properties/foo/properties/bar'));
    t.is(result, undefined);
});
ava_1.default('resolve $ref', t => {
    const schema = {
        definitions: {
            foo: {
                type: 'string'
            }
        },
        type: 'object',
        properties: {
            foos: {
                type: 'array',
                items: {
                    $ref: '#/definitions/foo'
                }
            }
        }
    };
    const result = util_1.Resolve.schema(schema, '#/properties/foos/items');
    t.deepEqual(result, { type: 'string' });
});
ava_1.default.failing('resolve $ref simple', t => {
    const schema = {
        definitions: {
            foo: {
                type: 'object',
                properties: {
                    bar: {
                        type: 'array',
                        items: {
                            $ref: '#/definitions/foo'
                        }
                    }
                }
            }
        },
        type: 'object',
        properties: {
            foos: {
                type: 'array',
                items: {
                    $ref: '#/definitions/foo'
                }
            }
        }
    };
    const result = util_1.Resolve.schema(schema, '#/properties/foos/items');
    t.deepEqual(result, {
        type: 'object',
        properties: {
            bar: {
                type: 'array',
                items: {
                    $ref: '#'
                }
            }
        }
    });
    t.not(schema.definitions.foo.properties.bar.items.$ref, '#');
});
ava_1.default.failing('resolve $ref complicated', t => {
    const schema = {
        definitions: {
            foo: {
                type: 'object',
                properties: {
                    bar: {
                        type: 'array',
                        items: {
                            $ref: '#/definitions/foo2'
                        }
                    }
                }
            },
            foo2: {
                type: 'object',
                properties: {
                    bar: {
                        type: 'array',
                        items: {
                            $ref: '#/definitions/foo'
                        }
                    }
                }
            }
        },
        type: 'object',
        properties: {
            foos: {
                type: 'array',
                items: {
                    $ref: '#/definitions/foo'
                }
            }
        }
    };
    const result = util_1.Resolve.schema(schema, '#/properties/foos/items');
    t.deepEqual(result, {
        definitions: {
            foo2: {
                type: 'object',
                properties: {
                    bar: {
                        type: 'array',
                        items: {
                            $ref: '#'
                        }
                    }
                }
            }
        },
        type: 'object',
        properties: {
            bar: {
                type: 'array',
                items: {
                    $ref: '#/definitions/foo2'
                }
            }
        }
    });
});
