// @jsonforms/ionic-renderers/lib/controls/object/object.control.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
/*
  The MIT License

  Copyright (c) 2018 EclipseSource Munich
  https://github.com/eclipsesource/jsonforms

  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:

  The above copyright notice and this permission notice shall be included in
  all copies or substantial portions of the Software.

  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
  THE SOFTWARE.
*/
var store_1 = require("@angular-redux/store");
var core_1 = require("@angular/core");
var angular_1 = require("@jsonforms/angular");
var core_2 = require("@jsonforms/core");
var ObjectControlRenderer = /** @class */ (function (_super) {
    __extends(ObjectControlRenderer, _super);
    function ObjectControlRenderer(ngRedux) {
        return _super.call(this, ngRedux) || this;
    }
    ObjectControlRenderer.prototype.mapAdditionalProps = function (props) {
        this.propsPath = props.path;
        this.detailUiSchema = props.findUISchema(props.scopedSchema, undefined, props.path);
    };
    ObjectControlRenderer = __decorate([
        core_1.Component({
            selector: 'jsonforms-object-control',
            template: "\n    <ion-card>\n      <ion-card-content>\n        <jsonforms-outlet\n          [uischema]=\"detailUiSchema\"\n          [schema]=\"scopedSchema\"\n          [path]=\"propsPath\"\n        >\n        </jsonforms-outlet>\n      </ion-card-content>\n    </ion-card>\n  "
        }),
        __metadata("design:paramtypes", [store_1.NgRedux])
    ], ObjectControlRenderer);
    return ObjectControlRenderer;
}(angular_1.JsonFormsControl));
exports.ObjectControlRenderer = ObjectControlRenderer;
exports.objectControlTester = core_2.rankWith(2, core_2.isObjectControl);
