// @jsonforms/ionic-renderers/lib/layouts/categorization/category/category.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var _ = __importStar(require("lodash"));
var ionic_angular_1 = require("ionic-angular");
var angular_1 = require("@jsonforms/angular");
var ParamsService_1 = require("../../../services/ParamsService");
var CategoryRenderer = /** @class */ (function (_super) {
    __extends(CategoryRenderer, _super);
    function CategoryRenderer(navParams, paramsService) {
        var _this = _super.call(this) || this;
        _this.paramsService = paramsService;
        var _a = navParams.get('category'), label = _a.label, uischema = _a.uischema, schema = _a.schema, path = _a.path;
        _this.label = label;
        _this.elements = uischema.elements.map(function (el) { return ({
            uischema: el,
            schema: schema,
            path: path
        }); });
        return _this;
    }
    CategoryRenderer.prototype.updatedElements = function () {
        var _this = this;
        var key = "jsonforms-category-" + this.label + "-" + this.path;
        var value = this.paramsService.get(key);
        if (value === undefined) {
            return this.elements;
        }
        this.paramsService.remove(key);
        this.elements = value.uischema.elements.map(function (el, index) {
            var existingEl = _this.elements[index];
            if (_.isEqual(existingEl.uischema, el)) {
                return existingEl;
            }
            return {
                uischema: el,
                schema: value.schema,
                path: value.path
            };
        });
        var remaining = value.uischema.elements.slice(this.elements.length);
        remaining.forEach(this.elements.push);
        return this.elements;
    };
    CategoryRenderer.CATEGORY_KEY = 'jsonforms-category-';
    CategoryRenderer = __decorate([
        core_1.Component({
            selector: 'jsonforms-category',
            template: "\n    <div *ngFor=\"let element of updatedElements()\">\n      <jsonforms-outlet [renderProps]=\"element\"></jsonforms-outlet>\n    </div>\n  "
        }),
        __metadata("design:paramtypes", [ionic_angular_1.NavParams, ParamsService_1.ParamsService])
    ], CategoryRenderer);
    return CategoryRenderer;
}(angular_1.JsonFormsBaseRenderer));
exports.CategoryRenderer = CategoryRenderer;
