// @jsonforms/angular/dist/test/src/control.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@jsonforms/core");
const core_2 = require("@angular/core");
const forms_1 = require("@angular/forms");
const base_renderer_1 = require("./base.renderer");
class JsonFormsControl extends base_renderer_1.JsonFormsBaseRenderer {
    constructor(ngRedux) {
        super();
        this.ngRedux = ngRedux;
        this.getEventValue = (event) => event.value;
        this.validator = (_c) => {
            return this.error ? { error: this.error } : null;
        };
        this.form = new forms_1.FormControl({
            value: '',
            disabled: true
        }, {
            updateOn: 'change',
            validators: this.validator.bind(this)
        });
    }
    onChange(ev) {
        const path = core_1.composeWithUi(this.uischema, this.path);
        this.ngRedux.dispatch(core_1.Actions.update(path, () => this.getEventValue(ev)));
        this.triggerValidation();
    }
    ngOnInit() {
        this.subscription = this.ngRedux
            .select()
            .subscribe((state) => {
            const props = this.mapToProps(state);
            const { data, enabled, errors, label, required, schema, uischema, visible } = props;
            this.label = core_1.computeLabel(core_1.isPlainLabel(label) ? label : label.default, required);
            this.data = data;
            this.error = errors ? errors.join('\n') : null;
            this.enabled = enabled;
            this.enabled ? this.form.enable() : this.form.disable();
            this.hidden = !visible;
            this.scopedSchema = core_1.Resolve.schema(schema, uischema.scope);
            this.description =
                this.scopedSchema !== undefined ? this.scopedSchema.description : '';
            this.id = props.id;
            this.form.setValue(data);
            this.mapAdditionalProps(props);
        });
        this.triggerValidation();
    }
    // @ts-ignore
    mapAdditionalProps(props) {
        // do nothing by default
    }
    ngOnDestroy() {
        if (this.subscription) {
            this.subscription.unsubscribe();
        }
    }
    getOwnProps() {
        const props = {
            uischema: this.uischema,
            schema: this.schema,
            path: this.path,
            id: this.id
        };
        if (this.disabled !== undefined) {
            props.enabled = !this.disabled;
        }
        if (this.visible !== undefined) {
            props.visible = this.visible;
        }
        return props;
    }
    mapToProps(state) {
        const props = core_1.mapStateToControlProps(state, this.getOwnProps());
        const dispatch = core_1.mapDispatchToControlProps(this.ngRedux.dispatch);
        return Object.assign({}, props, dispatch);
    }
    triggerValidation() {
        // these cause the correct update of the error underline, seems to be
        // related to ionic-team/ionic#11640
        this.form.markAsTouched();
        this.form.updateValueAndValidity();
    }
}
__decorate([
    core_2.Input(),
    __metadata("design:type", String)
], JsonFormsControl.prototype, "id", void 0);
__decorate([
    core_2.Input(),
    __metadata("design:type", Boolean)
], JsonFormsControl.prototype, "disabled", void 0);
__decorate([
    core_2.Input(),
    __metadata("design:type", Boolean)
], JsonFormsControl.prototype, "visible", void 0);
exports.JsonFormsControl = JsonFormsControl;
