// @jsonforms/ionic-renderers/lib/other/list-with-detail/pages/master/master.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var ionic_angular_1 = require("ionic-angular");
var AbstractMasterPage_1 = require("../AbstractMasterPage");
var core_2 = require("@jsonforms/core");
var store_1 = require("@angular-redux/store");
var MasterPage = /** @class */ (function (_super) {
    __extends(MasterPage, _super);
    function MasterPage(navParams, ngRedux) {
        var _this = _super.call(this) || this;
        _this.navParams = navParams;
        _this.ngRedux = ngRedux;
        _this.items = navParams.get('items');
        _this.schema = _this.navParams.get('schema');
        _this.uischema = _this.navParams.get('uischema');
        _this.path = _this.navParams.get('path');
        _this.pushDetail = _this.navParams.get('pushDetail');
        return _this;
    }
    MasterPage.prototype.ngOnInit = function () {
        var addItem = core_2.mapDispatchToArrayControlProps(this.ngRedux.dispatch, {
            uischema: this.uischema,
            schema: this.schema
        }).addItem;
        this.addItem = addItem;
    };
    MasterPage.prototype.onItemSelected = function (item) {
        this.pushDetail(item);
    };
    MasterPage.prototype.onClick = function () {
        this.addItem(this.path)();
    };
    MasterPage = __decorate([
        ionic_angular_1.IonicPage(),
        core_1.Component({
            selector: 'jsonforms-master-detail-master',
            template: "\n    <ion-content>\n      <ion-list>\n        <button\n          ion-item\n          *ngFor=\"let item of items\"\n          (click)=\"onItemSelected(item)\"\n        >\n          <h2>{{ item.label }}</h2>\n        </button>\n      </ion-list>\n      <ion-fab right bottom>\n        <button ion-fab (click)=\"onClick()\">\n          <ion-icon name=\"add\"></ion-icon>\n        </button>\n      </ion-fab>\n    </ion-content>\n  "
        }),
        __metadata("design:paramtypes", [ionic_angular_1.NavParams,
            store_1.NgRedux])
    ], MasterPage);
    return MasterPage;
}(AbstractMasterPage_1.AbstractMasterPage));
exports.MasterPage = MasterPage;
