// @jsonforms/core/dist/test/test/reducers/uischemas.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const ava_1 = __importDefault(require("ava"));
const _ = __importStar(require("lodash"));
const redux_mock_store_1 = __importDefault(require("redux-mock-store"));
const uischemas_1 = require("../../src/reducers/uischemas");
const actions_1 = require("../../src/actions");
const reducers_1 = require("../../src/reducers");
const generators_1 = require("../../src/generators");
ava_1.default('init state empty', t => {
    const dummyAction = {
        type: 'jsonforms/REMOVE_UI_SCHEMA',
        tester: undefined
    };
    t.deepEqual(uischemas_1.uischemaRegistryReducer(undefined, dummyAction), []);
});
ava_1.default('add ui schema', t => {
    const tester = () => 1;
    const control = {
        type: 'Control',
        scope: '#/definitions/foo'
    };
    const after = uischemas_1.uischemaRegistryReducer(undefined, actions_1.registerUISchema(tester, control));
    t.is(after.length, 1);
});
ava_1.default('remove ui schema', t => {
    const tester = () => 1;
    const control = {
        type: 'Control',
        scope: '#/definitions/foo'
    };
    const after = uischemas_1.uischemaRegistryReducer([
        {
            tester,
            uischema: control
        }
    ], actions_1.unregisterUISchema(tester));
    t.is(after.length, 0);
});
ava_1.default('findMatchingUISchema', t => {
    const testerA = (_schema, schemaPath) => _.endsWith(schemaPath, 'foo') ? 1 : 0;
    const testerB = (_schema, schemaPath) => _.endsWith(schemaPath, 'bar') ? 1 : 0;
    const controlA = {
        type: 'Control',
        scope: '#/definitions/foo'
    };
    const controlB = {
        type: 'Control',
        scope: '#/definitions/bar'
    };
    const before = [
        {
            tester: testerA,
            uischema: controlA
        },
        {
            tester: testerB,
            uischema: controlB
        }
    ];
    t.deepEqual(uischemas_1.findMatchingUISchema(before)(undefined, '#/defintions/foo', undefined), controlA);
    t.deepEqual(uischemas_1.findMatchingUISchema(before)(undefined, '#/defintions/bar', undefined), controlB);
});
ava_1.default('findUISchema returns generated UI schema if no match has been found', t => {
    const middlewares = [];
    const mockStore = redux_mock_store_1.default(middlewares);
    const store = mockStore({
        jsonforms: {
            core: {
                schema: {
                    definitions: {
                        baz: {
                            type: 'number'
                        }
                    }
                },
                data: undefined,
                uischema: undefined
            },
            uischemas: []
        }
    });
    t.deepEqual(reducers_1.findUISchema(store.getState())(reducers_1.getSchema(store.getState()), '#/definitions/baz', undefined), generators_1.Generate.uiSchema(reducers_1.getSchema(store.getState())));
});
ava_1.default('findMatchingUISchema with highest priority', t => {
    const testerA = (_schema, schemaPath) => _.endsWith(schemaPath, 'foo') ? 2 : 0;
    const testerB = (_schema, schemaPath) => _.endsWith(schemaPath, 'foo') ? 1 : 0;
    const controlA = {
        type: 'Control',
        scope: '#/definitions/foo'
    };
    const controlB = {
        type: 'Control',
        scope: '#/definitions/foo'
    };
    const before = [
        {
            tester: testerA,
            uischema: controlA
        },
        {
            tester: testerB,
            uischema: controlB
        }
    ];
    t.deepEqual(uischemas_1.findMatchingUISchema(before)(undefined, '#/definitions/foo', undefined), controlA);
});
